### PR TITLE
Add help for pkeyopt values for the genpkey commandline app.

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
Example output for
````
openssl genpkey -algorithm DHX -help
````

````
Usage: genpkey [options]

General options:
 -help               Display this summary
 -engine val         Use engine, possibly a hardware device
 -paramfile infile   Parameters file
 -algorithm val      The public key algorithm
 -verbose            Output status while generating keys
 -quiet              Do not output status while generating keys
 -pkeyopt val        Set the public key algorithm option as opt:value
 -config infile      Load a configuration file (this may load modules)

Output options:
 -out outfile        Output file
 -outform PEM|DER    output format (DER or PEM)
 -pass val           Output file pass phrase source
 -genparam           Generate parameters, not key
 -text               Print the in text
 -*                  Cipher to use to encrypt the key

Provider options:
 -provider-path val  Provider load path (must be before 'provider' argument if required)
 -provider val       Provider to load (can be specified multiple times)
 -propquery val      Property query used when fetching algorithms
Order of options may be important!  See the documentation.

pkeyopt values are
    type:string
    group:string
    priv_len:int
    pbits:uint
    qbits:uint
    digest:string
    properties:string
    gindex:int
    hexseed:string
    pcounter:int
    hindex:int

````

File: apps/genpkey.c [hunk 1]
```diff
@@ -57,6 +57,50 @@ const OPTIONS genpkey_options[] = {
...
+    if (ctx == NULL)
+        return;
+
+    if (EVP_PKEY_keygen_init(ctx) <= 0)
+        goto cleanup;
+    params = EVP_PKEY_CTX_settable_params(ctx);
+    if (params == NULL)
+        goto cleanup;
+
+    BIO_printf(bio_err, "\npkeyopt values are\n");
```
Comment (t8m) [inline apps/genpkey.c:93]: `\nThe possible -pkeyopt arguments are:\n`

File: doc/man1/openssl-genpkey.pod.in [hunk 2]
```diff
@@ -92,6 +92,9 @@ options supported depends on the public key algorithm used and its
 implementation. See L</KEY GENERATION OPTIONS> and
 L</PARAMETER GENERATION OPTIONS> below for more details.
 
+To list possible values for an algorithm use:
```
Comment (t8m) [inline doc/man1/openssl-genpkey.pod.in:95]: `possible I<opt> values`

**Review by t8m — APPROVED** [review]

File: apps/genpkey.c [hunk 3]
```diff
@@ -57,6 +57,50 @@ const OPTIONS genpkey_options[] = {
...
+{
+    EVP_PKEY_CTX *ctx = NULL;
+    const OSSL_PARAM *params;
+    int i, ishex = 0;
+
+    if (algname == NULL)
+        return;
+    ctx = EVP_PKEY_CTX_new_from_name(libctx, algname, propq);
+    if (ctx == NULL)
+        return;
```
Comment (hlandau) [inline apps/genpkey.c:85]: Should we not have some kind of output in the event of error (or at least a process exit code)?

File: apps/genpkey.c [hunk 4]
```diff
@@ -57,6 +57,50 @@ const OPTIONS genpkey_options[] = {
...
+    params = EVP_PKEY_CTX_settable_params(ctx);
+    if (params == NULL)
+        goto cleanup;
+
+    BIO_printf(bio_err, "\nThe possible -pkeyopt arguments are:\n");
+    for (i = 0; params[i].key != NULL; ++i) {
+        const char *name = param_datatype_2name(params[i].data_type, &ishex);
+
+        if (name != NULL)
+            BIO_printf(bio_err, "    %s%s:%s\n", ishex ? "hex" : "", params[i].key, name);
```
Comment (hlandau) [inline apps/genpkey.c:98]: This would print e.g. `hexpropname:string`... this seems confusing?

Perhaps `propname:string (hex)` / `propname:string`?

**Review by hlandau — CHANGES_REQUESTED** [review]

File: test/recipes/15-test_genpkey.t [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+push @algs, qw(DH DHX) unless disabled("dh");
+push @algs, qw(EC X25519 X448) unless disabled("ec");
+push @algs, qw(SM2) unless disabled("sm2");
+
+plan tests => scalar(@algs);
+
+foreach (@algs) {
+    my $alg = $_;
+
+    ok(run(app([ 'openssl', 'genpkey', '-algorithm', $alg, '-help'])),
```
Comment (paulidale) [inline test/recipes/15-test_genpkey.t:29]: Should the output be checked for minimal sanity?

E.g. the string "pkeyopt values are" appears.  Perhaps an option as well.

**Comment by levitte** [comment]
You do know, I hope, that you're kinda sorta reinventing the wheel?  Have a look at [`apps/include/app_params.h`](https://github.com/openssl/openssl/blob/master/apps/include/app_params.h) and [`apps/lib/app_params.c`](https://github.com/openssl/openssl/blob/master/apps/lib/app_params.c).

Of course, the output from those functions is bit more verbose, but I suspect it wouldn't be too hard to add a verbosity parameter.

Also, I would expect that this sort of help will find its way to all the other commands that use these sorts of opts too...  not asking for it in this PR, but I'd really like it if that was the intent, at least.

**Comment by slontis** [comment]
https://www.google.com/search?q=reinventing+the+wheel&client=firefox-b-e&sxsrf=AJOqlzWirg2zqy7PI3zZdrknVrMGFBlfuA:1673422900600&source=lnms&tbm=isch&sa=X&ved=2ahUKEwjIlLKMgr_8AhUMwjgGHUSVClcQ_AUoAXoECAIQAw&biw=1378&bih=619&dpr=2#imgrc=ZcomoXmbW4dgdM&imgdii=Ju5m33rWyILzZM

File: apps/genpkey.c (hunk 4 above)
Comment (slontis) [reply apps/genpkey.c:98]: I think it still needs the hex in front of the name as that is what you do to parse hex format.

**Comment by slontis** [comment]
> You do know, I hope, that you're kinda sorta reinventing the wheel?.

This is NOT trying to print out every possible param value like the list does.. Only ones that are actually usable by the command (e.g. OSSL_PARAM_UTF8_PTR is not something you can set from here).

**Comment by slontis** [comment]
> Also, I would expect that this sort of help will find its way to all the other commands that use these sorts of opts too... not asking for it in this PR, but I'd really like it if that was the intent, at least.

Is there really that many of them? This particular one is the one I keep on looking up all the time.

**Comment by levitte** [comment]
> > Also, I would expect that this sort of help will find its way to all the other commands that use these sorts of opts too... not asking for it in this PR, but I'd really like it if that was the intent, at least.
> 
> Is there really that many of them? This particular one is the one I keep on looking up all the time.

I haven't looked thoroughly, but I suspect that all `pkey` commands (`openssl pkeyutl`, `openssl pkeyparam`, `openssl pkey`) do, and I'm pretty sure that the newer operation commands (`openssl mac`, `openssl kdf`) do.  Also, for example `openssl dgst` has things like `-sigopt` and `-macopt`, although I'm unsure how relevant that is.

**Comment by paulidale** [comment]
Having the command I want to run produce a list is useful IMO. 

Knowing that `list` can also produce such (but in a less convenient verbosity) isn't a totally natural step.  It ought to be but it isn't.

**Review by paulidale — APPROVED** [review]

**Comment by tmshort** [comment]
@hlandau ?

**Comment by tmshort** [comment]
As @hlandau removed their stale review, this has passed the 24 hour threshold.

**Comment by tmshort** [comment]
Merged to master. Thank you for your contribution!

**Pull request closed without merging**